modules/apigee_edge_apidocs/src/ApiDocHtmlRouteProvider.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+/**
+ * Copyright 2019 Google Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+namespace Drupal\apigee_edge_apidocs;
+
+use Drupal\apigee_edge_apidocs\Form\ApiDocSettingsForm;
+use Drupal\Core\Entity\EntityTypeInterface;
+use Drupal\Core\Entity\Routing\AdminHtmlRouteProvider;
+use Symfony\Component\Routing\Route;
+
+/**
+ * Provides routes for API Doc entities.
+ *
+ * @see \Drupal\Core\Entity\Routing\AdminHtmlRouteProvider
+ * @see \Drupal\Core\Entity\Routing\DefaultHtmlRouteProvider
+ */
+class ApiDocHtmlRouteProvider extends AdminHtmlRouteProvider {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getRoutes(EntityTypeInterface $entity_type) {
+    $collection = parent::getRoutes($entity_type);
+
+    $entity_type_id = $entity_type->id();
+
+    if ($settings_form_route = $this->getSettingsFormRoute($entity_type)) {
+      $collection->add("$entity_type_id.settings", $settings_form_route);
+    }
+
+    if ($apidoc_collection_route = $collection->get('entity.apidoc.collection')) {
+      $apidoc_collection_route->setDefault('_title', 'API Docs');
+    }
+
+    return $collection;
+  }
+
+  /**
+   * Gets the settings form route.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeInterface $entity_type
+   *   The entity type.
+   *
+   * @return \Symfony\Component\Routing\Route|null
+   *   The generated route, if available.
+   */
+  protected function getSettingsFormRoute(EntityTypeInterface $entity_type) {
+    if (!$entity_type->getBundleEntityType()) {
+      $route = new Route("/admin/config/{$entity_type->id()}/settings");
+      $route
+        ->setDefaults([
+          '_form' => 'Drupal\apigee_edge_apidocs\Form\ApiDocSettingsForm',
+          '_title_callback' => ApiDocSettingsForm::class . '::title',
```

Comment:
Based on our experiences, it is not a good practice to use a form builder class method as a title callback because if you do that way the form class must implement the `ContainerInjectionInterface` if you would like to use dependency injected services in the title callback. At this moment, this requirement is fulfilled because FormBase, the current parent class of ApiDocSettingsForm implements this interface, but this can change in the future.
So for maintainability, it is better to create and use dedicated title provider classes, just like Drupal core does:  
https://github.com/drupal/core/blob/17557c97a816bc78e70989a86064c562bc27027f/lib/Drupal/Core/Entity/Controller/EntityController.php#L32
https://github.com/drupal/core/blob/17557c97a816bc78e70989a86064c562bc27027f/lib/Drupal/Core/Entity/Routing/DefaultHtmlRouteProvider.php#L120

Reply:
👍

Reply:
Resolved by dawehner: `'_title_callback' => ApiDocController::class . '::title',`